### PR TITLE
docs(api): new Moving Labware page

### PR DESCRIPTION
**Pull request by ecormany**
# Overview

A new page about Moving Labware, with the gripper and manually. It covers:

- Basic movement with `move_labware()`
- The differences between automatic and manual moves
- Which labware we officially support
- Moving onto and off of modules/adapters
- Moving onto and off of the deck with `OFF_DECK`

Also updates the API Reference entry for `move_labware()`.

Addresses RTC-145.

# Test Plan

- Code snippets tested against dev robot server on this branch, which rebased on `edge` on July 31.
- Code formatted with `black` using 80-character line
- Formatting and links tested in the [sandbox](http://sandbox.docs.opentrons.com/docs-moving-labware/v2/moving_labware.html)

# Changelog

- mention on Pipettes page that you don't have to `load_instrument` the gripper + crossref
- entirely new Moving Labware page
- uncomment `move_labware` and `OFF_DECK` from reference
- remove "this method may change" warning in `move_labware` docstring (it's time to commit to its functionality)
- [x] error type for trying to use gripper on OT-2
- removes old LPC parameters from reference
- general prose edits to reference

# Review requests

Please check all my Python code and make sure that it runs today (and ought to run on release PAPI 2.15). Confirm that we can remove docstring warnings and finalize `move_labware()` behavior.

# Risk assessment

low, docs only

### CODE REVIEW

File: api/docs/v2/moving_labware.rst [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            "opentrons_96_tiprack_1000ul", protocol_api.OFF_DECK
+        )
+        pipette = protocol.load_instrument(
+            "p1000_single_gen2", "left", tip_racks=[tips1, tips2]
+        )
+        # use all the on-deck tips
+        for i in range(96):
+            pipette.pick_up_tip()
+            pipette.drop_tip()
+        # move the spent tip rack off-deck
```
Comment (laviera) [inline api/docs/v2/moving_labware.rst]: Should this say "pause to move ..." as in the manual examples above - to make it clear this is what will happen?

File: api/docs/v2/moving_labware.rst [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        protocol.move_labware(labware=tips1, new_location=protocol_api.OFF_DECK)
+        # move the fresh tip rack on-deck
```
Comment (laviera) [inline api/docs/v2/moving_labware.rst]: Same comment - Should this say "pause to move ..." as in the manual move examples above ?

**Review by laviera — APPROVED** [review]

File: api/docs/v2/moving_labware.rst [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+.. code-block:: python
+    :substitutions:
+        
+    def run(protocol: protocol_api.ProtocolContext):
+        plate = protocol.load_labware('nest_96_wellplate_200ul_flat', 'D1')
+        protocol.move_labware(labware=plate, new_location='D2')
+        
+.. versionadded:: 2.15
+
+The required arguments of ``move_labware()`` are the ``labware`` you want to move and its ``new_location``. You don't need to specify where the move begins, since that information is already stored in the :py:class:`~opentrons.protocol_api.labware.Labware` object — ``plate`` in this example. The destination of the move can be any empty deck slot, or a module that's ready to have labware added to it (see :ref:`movement-modules` below). Attempting to move to an occupied location will raise a ``LocationIsOccupiedError``.
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst]: The last sentence in this paragraph:

> Attempting to move to an occupied location will raise a ``LocationIsOccupiedError``.

The preposition "to ... to" might be too much. Instead, could you try these?

- Any attempt to place labware in an occupied location will raise a ``LocationIsOccupiedError``.
- Attempting to place labware in an occupied location will raise a ``LocationIsOccupiedError``.

File: api/docs/v2/moving_labware.rst [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+.. note::
+    The slot that a labware currently occupies is considered an occupied destination. This prevents you from creating trivial moves, e.g., from slot D2 to slot D2.
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst]: The information in the note seems redundant (circular?) or perhaps too obvious to warrant the importance of a note call-out box. 

Maybe rephrase if this is an important point. Here's a suggestion:  

- The API won't let you move labware into the space it currently occupies. For example, if you already have a well plate in slot D2, you can't use the API to move it into the same slot (e.g., from slot D2 to slot D2).

File: api/docs/v2/moving_labware.rst [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+Automatic vs Manual Moves
+=========================
+
+There are two ways to move labware:
+
+- Automatically, with the Opentrons Flex Gripper.
+- Manually, by pausing the protocol until a user confirms that they've moved the labware.
+
+The ``use_gripper`` parameter of :py:meth:`~.ProtocolContext.move_labware` determines which way a particular movement should happen. Set its value to ``True`` for an automatic move. The default value is ``False``, so if you don't specify a value, the protocol will pause for a manual move.
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst]: Instead of "determines which way a particular movement should happen" let's call back to the 2 types of movement, like this:

- The ``use_gripper`` parameter of :py:meth:`~.ProtocolContext.move_labware` determines if a movement is automatic or manual.

File: api/docs/v2/moving_labware.rst [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+If you attempt to use the gripper to move labware in an OT-2 protocol, the API will raise a ``ErrorType TK`` error.
+
+
+Supported Labware
+=================
+
+You can manually move any standard or custom labware. Using the gripper to move the following types of labware is fully supported by Opentrons:
+
+.. list-table::
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst:74]: Love this list table with the bullets.  Teh awesomez.

File: api/docs/v2/moving_labware.rst [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+    Labware definitions don't explicitly declare compatibility or incompatibility with the gripper. The Python Protocol API won't raise a warning or error if you try to grip and move other types of labware.
+
+
+.. _movement-modules: 
+
+Movement with Modules
+=====================
+
+Moving labware on and off of modules lets you precisely control when the labware is in contact with the hot, cold, or magnetic surfaces of the modules — all within a single protocol.
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst:106]: Nice sentence. Had to note because of the way to echoes the features of various modules.

File: api/docs/v2/moving_labware.rst [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        hs_mod = protocol.load_module("heaterShakerModuleV1", "C1")
+        hs_adapter = hs_mod.load_adapter("opentrons_96_flat_bottom_adapter")
+        hs_mod.open_labware_latch()
+        protocol.move_labware(
+            labware=plate, new_location=hs_adapter, use_gripper=True
+        )
+
+.. versionadded:: 2.15
+
+If you try to move the plate to slot C1 or the Heater-Shaker module, you will get a ``LocationIsOccupiedError`` — C1 is occupied by the Heater-Shaker, and the Heater-Shaker is occupied by the adapter. Only the adapter, as the topmost item in that stack, is unoccupied.
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst]: I wonder if an image or illustration would be helpful here.

File: api/docs/v2/moving_labware.rst [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+In addition to moving labware around the deck, :py:meth:`~.ProtocolContext.move_labware` can also prompt you to move labware off of or onto the deck. 
+
+Remove labware from the deck to perform tasks like retrieving samples or discarding a spent tip rack. The destination location for such moves is the special constant :py:obj:`~opentrons.protocol_api.OFF_DECK`::
+
+    protocol.move_labware(labware=plate, new_location=protocol_api.OFF_DECK)
+    
+.. versionadded:: 2.15
+
+Moving labware off-deck always requires user intervention, since the gripper can't reach outside of the robot. Omit the ``use_gripper`` parameter or explicitly set it to ``False``. If you try to move labware off-deck with ``use_gripper=True``, the API will raise a ``LabwareMovementNotAllowedError``.
```
Comment (jwwojak) [inline api/docs/v2/moving_labware.rst]: Couple of thoughts for consideration here:

- Consider replacing "since" with "because."
- "Moving labware off-deck always requires user intervention because the gripper can't reach outside of the robot"

**Review by jwwojak — COMMENTED** [review]
@ecormany Added a few comments. As this deals with new API capabilities, perhaps other subject matter experts should make the make the merge go/no-go call. Mine mainly cover a few small text suggestions.

File: api/docs/v2/moving_labware.rst (hunk 4 above)
Comment (ecormany) [reply api/docs/v2/moving_labware.rst]: simplified this considerably.

File: api/src/opentrons/protocol_api/protocol_context.py
Comment (ecormany) [inline api/src/opentrons/protocol_api/protocol_context.py:1]: @sanni-t requesting your review on the docstrings, especially since this removes your TODO

File: api/docs/v2/moving_labware.rst [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+.. code-block:: python
+    :substitutions:
+        
+    def run(protocol: protocol_api.ProtocolContext):
+        plate = protocol.load_labware('nest_96_wellplate_200ul_flat', 'D1')
+        protocol.move_labware(labware=plate, new_location='D2')
+        
+.. versionadded:: 2.15
+
+The required arguments of ``move_labware()`` are the ``labware`` you want to move and its ``new_location``. You don't need to specify where the move begins, since that information is already stored in the :py:class:`~opentrons.protocol_api.labware.Labware` object — ``plate`` in this example. The destination of the move can be any empty deck slot, or a module that's ready to have labware added to it (see :ref:`movement-modules` below). Movement to an occupied location — including the labware's current location — will raise a ``LocationIsOccupiedError``.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst]: I don't think we should mention `LocationIsOccupiedError`. We can just say "raise an exception" instead.

-----

There *is* something called a `LocationIsOccupiedError`, and it *will* stop you from moving labware to an occupied location. But it's currently a deep internal thing. In practice, by the time it reaches the user's Python script, it's wrapped up and hidden behind something else. If somebody did `except LocationIsOccupiedError`, it wouldn't work.

Also, maybe more importantly, `LocationIsOccupiedError` is only importable through `opentrons.protocol_engine.errors`. That's not one of the existing officially-exposed modules (like `opentrons.protocol_api` and `opentrons.execute`). So even if it were visible in practice to the user's Python script, it wouldn't be officially sanctioned.

File: api/docs/v2/moving_labware.rst [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        protocol.move_labware(labware=plate, new_location='C1')
+
+.. versionadded:: 2.15
+
+.. note::
+    Don't add a ``pause()`` command before ``move_labware()``. When ``use_gripper`` is unset or ``False``, the protocol pauses when it reaches the movement step. The Opentrons App or the touchscreen on Flex shows an animation of the labware movement that you need to perform manually. The protocol only resumes when you press **Confirm and resume**.
+
+The above example is a complete and valid ``run()`` function. You don't have to load the gripper as an instrument, and there is no ``InstrumentContext`` for the gripper. All you have to do to specify that a protocol requires the gripper is to include at least one ``move_labware()`` command with ``use_labware=True``.
+
+If you attempt to use the gripper to move labware in an OT-2 protocol, the API will raise a ``NotSupportedOnRobotType`` error.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst]: See my other comment about not mentioning specific exception types.

File: api/docs/v2/moving_labware.rst [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+          * ``opentrons_flex_96_tiprack_1000ul``
+          * ``opentrons_flex_96_filtertiprack_50ul``
+          * ``opentrons_flex_96_filtertiprack_200ul``
+          * ``opentrons_flex_96_filtertiprack_1000ul``
+    
+The gripper may work with other ANSI/SLAS standard labware, but this is not recommended.
+
+.. note::
+
+    Labware definitions don't explicitly declare compatibility or incompatibility with the gripper. The Python Protocol API won't raise a warning or error if you try to grip and move other types of labware.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst:98]: @sanni-t Do you know if this is the strategy we'll have for release?

File: api/docs/v2/moving_labware.rst [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+.. versionadded:: 2.15
+
+If you try to move the plate to slot C1 or the Heater-Shaker module, you will get a ``LocationIsOccupiedError`` — C1 is occupied by the Heater-Shaker, and the Heater-Shaker is occupied by the adapter. Only the adapter, as the topmost item in that stack, is unoccupied.
+
+Also note the ``hs_mod.open_labware_latch()`` command in the above example. To move labware onto or off of a module, you have to make sure that it's physically accessible:
+
+    - For the Heater-Shaker, use :py:meth:`~.HeaterShakerContext.open_labware_latch`.
+    - For the Thermocycler, use :py:meth:`~.ThermocyclerContext.open_lid`.
+    
+If the labware is inaccessible, the API will raise a ``LabwareMovementNotAllowedError``.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst]: See my other comment about not mentioning specific exception types.

File: api/docs/v2/moving_labware.rst [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        hs_mod = protocol.load_module("heaterShakerModuleV1", "C1")
+        hs_adapter = hs_mod.load_adapter("opentrons_96_flat_bottom_adapter")
+        hs_mod.open_labware_latch()
+        protocol.move_labware(
+            labware=plate, new_location=hs_adapter, use_gripper=True
+        )
+
+.. versionadded:: 2.15
+
+If you try to move the plate to slot C1 or the Heater-Shaker module, you will get a ``LocationIsOccupiedError`` — C1 is occupied by the Heater-Shaker, and the Heater-Shaker is occupied by the adapter. Only the adapter, as the topmost item in that stack, is unoccupied.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst]: See my other comment about not mentioning specific exception types.

File: api/docs/v2/moving_labware.rst [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+In addition to moving labware around the deck, :py:meth:`~.ProtocolContext.move_labware` can also prompt you to move labware off of or onto the deck. 
+
+Remove labware from the deck to perform tasks like retrieving samples or discarding a spent tip rack. The destination location for such moves is the special constant :py:obj:`~opentrons.protocol_api.OFF_DECK`::
+
+    protocol.move_labware(labware=plate, new_location=protocol_api.OFF_DECK)
+    
+.. versionadded:: 2.15
+
+Moving labware off-deck always requires user intervention, because the gripper can't reach outside of the robot. Omit the ``use_gripper`` parameter or explicitly set it to ``False``. If you try to move labware off-deck with ``use_gripper=True``, the API will raise a ``LabwareMovementNotAllowedError``.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst]: See my other comment about not mentioning specific exception types.

File: api/docs/v2/moving_labware.rst [hunk 16]
```diff
@@ -0,0 +1,171 @@
+:og:description: How to move labware, with the Flex Gripper or manually, in a Python protocol.
+
+.. _moving-labware:
+
+**************
+Moving Labware
+**************
+
+You can move an entire labware (and all of its contents) from one deck slot to another at any point during your protocol. On Flex, you can either use the gripper or move the labware manually. On OT-2, you can can only move labware manually, since it doesn't have a gripper instrument.
```
Comment (SyntaxColoring) [inline api/docs/v2/moving_labware.rst:9]: Should we say "On OT-2" and "On Flex"? I usually say "On an OT-2" and "On a Flex".

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 17]
```diff
@@ -534,11 +534,8 @@ def move_labware(
     ) -> None:
         """Move a loaded labware to a new location.
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/protocol_context.py]: I think a cross-reference from here to moving_labware.rst would be helpful.

File: api/docs/v2/new_protocol_api.rst [hunk 18]
```diff
@@ -80,11 +76,8 @@ Useful Types and Definitions
 .. automodule:: opentrons.types
    :members: PipetteNotAttachedError, Point, Location, Mount
 
-..
-   TODO(mm, 2023-06-06): This should be uncommented when move_labware() is published. See above.
-   https://opentrons.atlassian.net/browse/RTC-288
-   .. autodata:: opentrons.protocol_api.OFF_DECK
-      :no-value:
+.. autodata:: opentrons.protocol_api.OFF_DECK
```
Comment (SyntaxColoring) [inline api/docs/v2/new_protocol_api.rst:79]: We might also want to edit the `opentrons.protocol_api.OFF_DECK` [docstring](https://github.com/Opentrons/opentrons/blob/aa1665a41fd68e244e9a12a0e1bf57b7613583d9/api/src/opentrons/protocol_api/_types.py#L14-L18) to reference your nice new `moving_labware.rst` overview page. It currently references the terser `move_labware()` reference docs.

File: api/src/opentrons/protocol_api/protocol_context.py [hunk 19]
```diff
@@ -549,27 +546,23 @@ def move_labware(
...
+        Gripper-only parameters:
 
-        :param use_pick_up_location_lpc_offset: Whether to use LPC offset of the labware
-                                                associated with its pick up location.
-        :param use_drop_location_lpc_offset: Whether to use LPC offset of the labware
-                                             associated with its drop off location.
-        :param pick_up_offset: Offset to use when picking up labware.
-        :param drop_offset: Offset to use when dropping off labware.
+        :param pick_up_offset: Optional x, y, z vector offset to use when picking up labware.
+        :param drop_offset: Optional x, y, z vector offset to use when dropping off labware.
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/protocol_context.py:560]: This is good for the purposes of this PR, just thinking for next steps:

The `{"x": 1, "y": 2, "z": 3}` shape of these dicts is, I think, unique to this function, so we might have to document it more explicitly.

Or, @sanni-t: should we turn these arguments into `opentrons.types.Point`s?

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_api/protocol_context.py:560]: I guess let's leave it as you have it in this PR, at least.

**Review by SyntaxColoring — COMMENTED** [review]
This looks great to me! One gotcha about mentioning specific exception types, though.

File: api/docs/v2/moving_labware.rst (hunk 16 above)
Comment (ecormany) [reply api/docs/v2/moving_labware.rst:9]: I don't think we have firm style guidance on this. My brain thinks of "on OT-2" meaning on the platform in general, and "on an OT-2" meaning on a particular (even if indeterminate) robot. 

But maybe I've just been Apple-pilled by Tim Cook saying "on iPhone", etc.

**Comment by ecormany** [comment]
We've had a lot of eyes and general 👍 on this PR. Going to merge now, and if there are any pre-release changes from #13222 or anywhere else, we can address them in a follow-up.
